include/alpaka/core/Assert.hpp

```diff
@@ -20,8 +20,15 @@
 
 #ifdef ALPAKA_DEBUG_OFFLOAD_ASSUME_HOST
 #    define ALPAKA_ASSERT_OFFLOAD(EXPRESSION) ALPAKA_ASSERT(EXPRESSION)
+#elif defined __AMDGCN__ && (!defined NDEBUG)
+#    define ALPAKA_ASSERT_OFFLOAD(EXPRESSION)                                                                         \
+        if(!(EXPRESSION))                                                                                             \
+            __builtin_trap();                                                                                         \
+        do                                                                                                            \
+        {                                                                                                             \
+        } while(false)
 #else
-#    define ALPAKA_ASSERT_OFFLOAD(EXPRESSION)
+#    define ALPAKA_ASSERT_OFFLOAD(EXPRESSION) alpaka::ignore_unused(EXPRESSION)
```

Comment:
This macro should be empty. Normal `assert` is also not performing anything if `NDEBUG` is defined.

IMO this macro should remove the expression if `NDEBUG` is defined and guards around  `ALPAKA_ASSERT_OFFLOAD()` must be removed.

Reply:
This change is the reason the why the `#ifndef NDEBUG` is required in some places: it leaves the expression in place to not trigger unused argument warnings. In exchange leaving the expression in place makes it so, that if the expression uses some identifier not declared when `defined NDEBUG` then the whole macro needs to be guarded.

Maybe the better solution would to not add `alpaka::ingore_unsused` and leave this case to the calling code to deal with. I think such a guard is in place for `ALPAKA_ASSERT`, so this would indeed be more consistent.

Reply:
Removing `alpaka::ingore_unused` breaks the debug build, because the existing guards for unused parameters assume, that the parameter will be used in debug mode, but with `ALPAKA_OFFLOAD_ASSUME_HOST=OFF` this would then not be the case, hence the definition as `alpaka::ingore_unused(EXPRESSION)`.

Reply:
The only way to reconcile this would be to have an unconditional `alpaka::ingore_unused(param)` at the affected cal sites... .